### PR TITLE
perf: enhance performance in state

### PR DESCRIPTION
**Pull request by faddat**
Working on #1976 and also examining other ways to improve performance. 

closes: #1976 

---

#### PR checklist

- [x] Tests written/updated
- [ ] Changelog entry added in `.changelog` (we use [unclog](https://github.com/informalsystems/unclog) to manage our changelog)
- [ ] Updated relevant documentation (`docs/` or `spec/`) and code comments

### CODE REVIEW

**Comment by faddat** [comment]
Needs #1954

**Comment by faddat** [comment]
@ValarDragon - I don't think that this is quite done correctly yet, but could you take a peek?

**Comment by faddat** [comment]
oh I def found a bug.... it is strange that tests pass

**Comment by faddat** [comment]
I was afraid of that...

**Comment by faddat** [comment]
@ValarDragon @melekes - what would be a good way to benchmark here?

And what should be "the standard"?

**Comment by faddat** [comment]
This isn't the same thing reported by @ValarDragon but I think it's related:

https://github.com/cometbft/cometbft-db/pull/112#issuecomment-1883788848

File: internal/state/store.go [hunk 1]
```diff
@@ -131,12 +139,26 @@ func IsEmpty(store dbStore) (bool, error) {
 
 // NewStore creates the dbStore of the state pkg.
 func NewStore(db dbm.DB, options StoreOptions) Store {
-	return dbStore{db, options}
+	return &dbStore{
+		db:              db,
+		valInfoCache:    make(map[int64]*cmtstate.ValidatorsInfo),
```
Comment (melekes) [inline internal/state/store.go:144]: Why do we need these? I don't see them used anywhere in this PR

**Review by melekes — COMMENTED** [review]
While there are some nice improvements here, this PR fails to address the main issue described in https://github.com/cometbft/cometbft/issues/1976

File: internal/state/store.go (hunk 1 above)
Comment (faddat) [reply internal/state/store.go:144]: I believe that I made a branch management mistake here.  Thanks for review.

File: internal/state/store.go [hunk 2]
```diff
@@ -403,20 +426,19 @@ func (store dbStore) PruneStates(from int64, to int64, evidenceThresholdHeight i
 		}
 		pruned++
 
-		// avoid batches growing too large by flushing to database regularly
-		if pruned%1000 == 0 && pruned > 0 {
-			err := batch.Write()
-			if err != nil {
+		// Commit the batch after a certain number of operations to avoid large batches
+		if pruned%batchSize == 0 {
+			if err := batch.WriteSync(); err != nil {
```
[the remark this answers is not in the document]
Comment (faddat) [reply internal/state/store.go:431]: let me see if batch.Write() works just as well.

**Comment by faddat** [comment]
@melekes I will do more here, once 

* https://github.com/cometbft/cometbft/pull/2021

has been merged.

File: internal/state/store.go (hunk 1 above)
Comment (faddat) [reply internal/state/store.go:144]: Hey, I'm wrong.  Basically one finding was that it was faster to use pointers and lay things out first.  That's why that's there.

**Comment by faddat** [comment]
I'm opening this for review again -- close if anyone thinks is not needed, it seems like I went outside of the original intended scope.

**Comment by faddat** [comment]
Hey I'm going to close this with the note to self that we should have more benchmarks across the code base

**Pull request closed without merging**